File: problem_4/python/problem4.py
#!/usr/bin/python

def int_to_list(number):
  return map(int, str(number))

def is_palindrome(number):
  local_list = int_to_list(number)
  return local_list == list(reversed(local_list))

if __name__ == "__main__":
  print "%s" % max([x * y for x in range(1000,100, -1) \
                          for  y  in range(1000,100,-1) \
                          if is_palindrome(x * y)])

File: problem_4/perl/problem4.pl
#!/usr/bin/perl

use strict;
use warnings;
use List::Util qw(max);

sub is_palimdrone
{
  my ($number) = @_;

  my @digits = split(//, $number);
  my @reversed_digits = reverse(@digits);

  # got this code idea from url below;  I've modified it slightly.
  # http://perl.active-venture.com/pod/perlfaq4-dataarrays.html
  for (my $i = 0; $i < @digits; $i++)
  {
    return 0 if $digits[$i] != $reversed_digits[$i];
  }

  return 1;
}

my @two = my @one = reverse((100..999));
my @three;

for my $one (@one)
{
  for my $two (@two)
  {
    my $three = $two * $one;
    next unless is_palimdrone($three);
    push @three, $three;
  }
}

print max(@three);
